src/js/components/modals/JobFormModal.js

```diff
@@ -123,60 +131,179 @@ class JobFormModal extends mixin(StoreMixin) {
   }
 
   resetState() {
+    let {job} = this.props;
     this.setState({
       defaultTab: '',
       errorMessage: null,
-      job: this.props.job,
+      jobFormModel: JobUtil.createFormModelFromSchema(JobSchema, job),
+      jobJsonString: JSON.stringify(cleanJobJSON(JobUtil.createJobSpecFromJob(job)), null, 2),
       jsonMode: false
     });
   }
 
+  createJobFromEditorContents(keepValidationErrors = false) {
+    let {jobJsonString, jsonMode} = this.state;
+
+    if (jsonMode) {
+
+      // Try to parse JSON string and detect errors
+      var jobDefinition = null;
+      try {
+        jobDefinition = JSON.parse(jobJsonString);
+      } catch (e) {
+        this.setState({
+          errorMessage: {
+            message: 'Invalid JSON syntax',
+            details: [
+              { path: '/', errors: [e.toString()] }
+            ]
+          }
+        });
+        return null;
+      }
+
+      // Really hackish way to validate the json string schema, trying to re-use
+      // as much code as possilbe without getting nasty.
+      let dummyItemRenderer = function () { return (<div></div>); };
+      let job = new Job(jobDefinition);
+      let formModel = JobUtil.createFormModelFromSchema(JobSchema, job);
+      let formMultiDef = SchemaUtil.schemaToMultipleDefinition({
+        schema: JobSchema,
+        renderSubheader: dummyItemRenderer,
+        renderLabel: dummyItemRenderer,
+        renderRemove: dummyItemRenderer,
+        renderAdd: dummyItemRenderer
+      });
+      let errorDetails = [];
+      FormUtil.forEachDefinition(formMultiDef, (definition) => {
+        definition.showError = false;
+
+        if (typeof definition.externalValidator !== 'function') {
+          return null;
+        }
+
+        let fieldValidated = definition.externalValidator(formModel, definition);
+        if (!fieldValidated) {
+          errorDetails.push({
+            path: '/', errors: [definition.showError]
+          });
+        }
+      });
+
+      // If we have errors, display them
+      if (errorDetails.length) {
+        this.setState({
+          errorMessage: {
+            message: 'There are errors in your JSON definition',
+            details: errorDetails
+          }
+        });
+        if (!keepValidationErrors) {
+          return null;
+        }
+      }
+
+      return job;
+
+    } else {
```

Comment:
The else is not needed since we are returning at the end of the if block in line 206.